In [3]:
# Commands to install some of the libraries in-case if they are not installed
# Any other library that needs to be installed just use: !pip install <library name>
    
!pip install seaborn
!pip install missingno
!pip install xgboost
!pip install catboost
!pip install regex
!pip install sklearn
!pip install pandas
!pip install numpy
!pip install imblearn
!pip install lightgbm

  error: subprocess-exited-with-error
  
  python setup.py egg_info did not run successfully.
  exit code: 1
  
  [15 lines of output]
  The 'sklearn' PyPI package is deprecated, use 'scikit-learn'
  rather than 'sklearn' for pip commands.
  
  Here is how to fix this error in the main use cases:
  - use 'pip install scikit-learn' rather than 'pip install sklearn'
  - replace 'sklearn' by 'scikit-learn' in your pip requirements files
    (requirements.txt, setup.py, setup.cfg, Pipfile, etc ...)
  - if the 'sklearn' package is used by one of your dependencies,
    it would be great if you take some time to track which package uses
    'sklearn' instead of 'scikit-learn' and report it to their issue tracker
  - as a last resort, set the environment variable
    SKLEARN_ALLOW_DEPRECATED_SKLEARN_PACKAGE_INSTALL=True to avoid this error
  
  More information is available at
  https://github.com/scikit-learn/sklearn-pypi-package
  [end of output]
  
  note: This error originates from a subpr


  Using cached sklearn-0.0.post12.tar.gz (2.6 kB)
  Preparing metadata (setup.py): started
  Preparing metadata (setup.py): finished with status 'error'


In [4]:
import pandas as pd   # data processing, CSV file I/O (e.g. pd.read_csv)
import numpy as np   # linear algebra
import matplotlib.pyplot as plt  #graphs and plots
import seaborn as sns   #data visualizations
import csv # Some extra functionalities for csv  files - reading it as a dictionary
from lightgbm import LGBMClassifier #sklearn is for machine learning and statistical modeling including classification, regression, clustering and dimensionality reduction

from sklearn.model_selection import train_test_split, cross_validate   #break up dataset into train and test sets

from sklearn.linear_model import LogisticRegression
from sklearn.ensemble import RandomForestClassifier
from sklearn.model_selection import cross_val_score
from sklearn.preprocessing import StandardScaler, MinMaxScaler


# importing python library for working with missing data
import missingno as msno
# To install missingno use: !pip install missingno
import re    # This library is used to perform regex pattern matching

# import various functions from sklearn
from sklearn.linear_model import LogisticRegression, LinearRegression
from sklearn.neighbors import KNeighborsClassifier
from sklearn.tree import DecisionTreeClassifier
from sklearn.ensemble import RandomForestClassifier
from sklearn.svm import LinearSVC
from sklearn.ensemble import GradientBoostingClassifier
from catboost import CatBoostClassifier
import xgboost as xgb
from sklearn.metrics import roc_auc_score, accuracy_score, precision_score, recall_score, classification_report, make_scorer
from sklearn.linear_model import SGDClassifier
from sklearn.model_selection import GridSearchCV, RandomizedSearchCV, train_test_split

Import additional items as needed...
We may not use them all in this course...

In [5]:
conda install -c conda-forge imbalanced-learn

Retrieving notices: ...working... done
Solving environment: ...working... done

# All requested packages already installed.


Note: you may need to restart the kernel to use updated packages.




==> WARNING: A newer version of conda exists. <==
  current version: 23.7.4
  latest version: 24.1.2

Please update conda by running

    $ conda update -n base -c defaults conda

Or to minimize the number of packages updated during conda update use

     conda install conda=24.1.2




In [4]:
from imblearn.over_sampling import SMOTE

In [5]:
from sklearn.model_selection import KFold,cross_val_score, RepeatedStratifiedKFold,StratifiedKFold
from sklearn.impute import SimpleImputer
from sklearn.pipeline import Pipeline
from sklearn.discriminant_analysis import LinearDiscriminantAnalysis
from sklearn.preprocessing import OneHotEncoder,StandardScaler,PowerTransformer
from sklearn.compose import ColumnTransformer
from sklearn.pipeline import Pipeline
from sklearn.impute import KNNImputer,SimpleImputer
from sklearn.compose import make_column_transformer
from imblearn.pipeline import make_pipeline
from sklearn.svm import SVC
from sklearn.impute import SimpleImputer
from sklearn.dummy import DummyClassifier
from imblearn.over_sampling import SMOTE
from sklearn.neighbors import KNeighborsClassifier
from sklearn.metrics import confusion_matrix, accuracy_score, balanced_accuracy_score,\
                            precision_score, recall_score, roc_auc_score,\
                            classification_report, f1_score

import plotly 
import plotly.express as px
import plotly.graph_objs as go
import plotly.offline as py
from plotly.offline import iplot
from plotly.subplots import make_subplots
import plotly.figure_factory as ff

import warnings
warnings.filterwarnings("ignore")

* If from imblearn.oversampling import SMOTE does not load use
    `conda install -c conda-forge imbalanced-learn`
* Then rerun
    `from imblearn.over_sampling import SMOTE`

In [7]:
diabetes_new = pd.read_csv("C:\\Users\\johnc\\Downloads\\diabetes_test.csv")


In [8]:
diabetes_new['readmitted'] = diabetes_new['readmitted'].replace({'NO':0,'>30':0,'<30':1})
print(diabetes_new)

                   race  gender      age  time_in_hospital  \
0             Caucasian  Female   [0-10)                 1   
1             Caucasian  Female  [10-20)                 3   
2       AfricanAmerican  Female  [20-30)                 2   
3             Caucasian    Male  [30-40)                 2   
4             Caucasian    Male  [40-50)                 1   
...                 ...     ...      ...               ...   
101761  AfricanAmerican    Male  [70-80)                 3   
101762  AfricanAmerican  Female  [80-90)                 5   
101763        Caucasian    Male  [70-80)                 1   
101764        Caucasian  Female  [80-90)                10   
101765        Caucasian    Male  [70-80)                 6   

        num_lab_procedures  num_procedures  num_medications  \
0                       41               0                1   
1                       59               0               18   
2                       11               5               13   
3  

# Breaking the data up into Train & Test

In [9]:
# Using get_dummies from the pandas library to convert categorical feature to numeric
diabetes_ready = pd.get_dummies(diabetes_new, drop_first = False)


In [10]:
print(diabetes_ready)

        time_in_hospital  num_lab_procedures  num_procedures  num_medications  \
0                      1                  41               0                1   
1                      3                  59               0               18   
2                      2                  11               5               13   
3                      2                  44               1               16   
4                      1                  51               0                8   
...                  ...                 ...             ...              ...   
101761                 3                  51               0               16   
101762                 5                  33               3               18   
101763                 1                  53               0                9   
101764                10                  45               2               21   
101765                 6                  13               3                3   

        number_outpatient  

In [11]:
train_df, valid_df, test_df = np.split(diabetes_ready.sample(frac=1, random_state=42),
                                       [int(.7*len(diabetes_ready)), int(0.85*len(diabetes_ready))])
train_df = train_df.reset_index(drop = True)
valid_df = valid_df.reset_index(drop = True)
test_df = test_df.reset_index(drop = True)

In [12]:
diabetes_ready.readmitted.value_counts()

readmitted
0    90409
1    11357
Name: count, dtype: int64

In [13]:
train_df.readmitted.value_counts()

readmitted
0    63275
1     7961
Name: count, dtype: int64

In [14]:
valid_df.readmitted.value_counts()

readmitted
0    13520
1     1745
Name: count, dtype: int64

In [15]:
test_df.readmitted.value_counts()

readmitted
0    13614
1     1651
Name: count, dtype: int64

# Treating the Imbalance in the Data

The data above contians an imbalance of data since there is significanlty more patients that were not readmitted comapred to those that were. 13K vs 1.6K. Thus I need to correct this imbalance by synthesizing data.

In [16]:
def calc_prevalence(y_actual):
    
    '''
    This function is to understand the ratio/distribution of the classes that we are going to predict for.
    
    Params:
    1. y_actual: The target feature
    
    Return:
    1. (sum(y_actual)/len(y_actual)): The ratio of the postive class in the comlpete data.
    '''
    
    return (sum(y_actual)/len(y_actual))

In [17]:
rows_pos = train_df.readmitted == 1
df_train_pos = train_df.loc[rows_pos]
df_train_neg = train_df.loc[~rows_pos]

diabetes_df_balanced = pd.concat([df_train_pos, df_train_neg.sample(n = len(df_train_pos), random_state = 111)],axis = 0)

diabetes_df_balanced = diabetes_df_balanced.sample(n = len(diabetes_df_balanced), random_state = 42).reset_index(drop = True)

print('Train balanced prevalence(n = %d):%.3f'%(len(diabetes_df_balanced), \
                                                calc_prevalence(diabetes_df_balanced.readmitted.values)))

Train balanced prevalence(n = 15922):0.500


Now after splitting the training data into postive and negative I was able to merge the balanced data and shuffle its order. As can be seen below both counts of the readmitted column are now balanced. 

In [18]:
diabetes_df_balanced.readmitted.value_counts()

readmitted
1    7961
0    7961
Name: count, dtype: int64

Below I am creating variables for me test, validation and training datasets to seperate the feature variables from the target variable.

In [19]:
X_train = diabetes_df_balanced.drop('readmitted',axis=1)

y_train = diabetes_df_balanced['readmitted']

X_valid = valid_df.drop('readmitted',axis=1)

y_valid = valid_df['readmitted']

X_test = test_df.drop('readmitted',axis=1)

y_test = test_df['readmitted']

In [20]:
scaler=StandardScaler()
X_train[['time_in_hospital','num_lab_procedures','num_procedures','num_medications','number_outpatient','number_emergency','number_inpatient','number_diagnoses']] = pd.DataFrame(scaler.fit_transform(X_train[['time_in_hospital','num_lab_procedures','num_procedures','num_medications','number_outpatient','number_emergency','number_inpatient','number_diagnoses']]),columns=['time_in_hospital','num_lab_procedures','num_procedures','num_medications','number_outpatient','number_emergency','number_inpatient','number_diagnoses'])
X_valid[['time_in_hospital','num_lab_procedures','num_procedures','num_medications','number_outpatient','number_emergency','number_inpatient','number_diagnoses']] = pd.DataFrame(scaler.transform(X_valid[['time_in_hospital','num_lab_procedures','num_procedures','num_medications','number_outpatient','number_emergency','number_inpatient','number_diagnoses']]),columns=['time_in_hospital','num_lab_procedures','num_procedures','num_medications','number_outpatient','number_emergency','number_inpatient','number_diagnoses'])
X_test[['time_in_hospital','num_lab_procedures','num_procedures','num_medications','number_outpatient','number_emergency','number_inpatient','number_diagnoses']] = pd.DataFrame(scaler.transform(X_test[['time_in_hospital','num_lab_procedures','num_procedures','num_medications','number_outpatient','number_emergency','number_inpatient','number_diagnoses']]),columns=['time_in_hospital','num_lab_procedures','num_procedures','num_medications','number_outpatient','number_emergency','number_inpatient','number_diagnoses'])

# Creating and Understanding Models

In [21]:
def calc_specificity(y_actual, y_pred, thresh):
    # calculates specificity
    return sum((y_pred < thresh) & (y_actual == 0)) /sum(y_actual ==0)

def print_report(y_actual, y_pred, thresh = 0.5):
    
    '''
    This function calculates all the metrics to asses the machine learning models.
    
    Params:
    1. y_actual: The actual values for the target variable.
    2. y_pred: The predicted values for the target variable.
    3. thresh: The threshold for the probability to be considered as a positive class. Default value 0.5
    
    Return:
    1. AUC
    2. Accuracy
    3. Recall
    4. Precision
    5. Specificity
    '''
    
    auc = roc_auc_score(y_actual, y_pred)
    accuracy = accuracy_score(y_actual, (y_pred > thresh))
    recall = recall_score(y_actual, (y_pred > thresh))
    precision = precision_score(y_actual, (y_pred > thresh))
    specificity = calc_specificity(y_actual, y_pred, thresh)
    print('AUC:%.3f'%auc)
    print('accuracy:%.3f'%accuracy)
    print('recall:%.3f'%recall)
    print('precision:%.3f'%precision)
    print('specificity:%.3f'%specificity)
    print('prevalence:%.3f'%calc_prevalence(y_actual))
    print(' ')
    return auc, accuracy, recall, precision, specificity

## Linear Regression

In [57]:
lnr = LinearRegression()
lnr.fit(X_train, y_train)


y_valid_preds = lnr.predict(X_valid)

In [58]:
y_valid_preds

array([0.48852539, 0.71655273, 0.67602539, ..., 0.47875977, 0.66186523,
       0.55004883])

## Logistic Regression

In [59]:
lr=LogisticRegression(random_state = 42, solver = 'newton-cg', max_iter = 200)
lr.fit(X_train, y_train)

y_valid_preds = lr.predict_proba(X_valid)[:,1]

print('Metrics for Validation data:')

lr_valid_auc, lr_valid_accuracy, lr_valid_recall, \
    lr_valid_precision, lr_valid_specificity = print_report(y_valid,y_valid_preds, 0.5)

Metrics for Validation data:
AUC:0.631
accuracy:0.653
recall:0.513
precision:0.168
specificity:0.671
prevalence:0.114
 


Accoring to the results of the logistic regression:
AUC: A score of 0.631 means that the model is able to do slightly better than just random chance.
Accuracy: 65.3% of the time it was able to classify the cases correctly in the validation set.
Recall: 51.3% of the time the model was able to correctly identify true positve cases from all postive cases in the validation set.
Precision: When the model predicts a postive case it only does so correctly 16.8% of the time.
Specificity: 67.1% of the time the model was able to correctly identify true negative cases from all negative cases in the validation set.
Prevalance: In the validation set only 11.4% of the cases were postive.

Overall it seems that this model isn't reliable to predict readmission of patients. I know that I would have to do some fine tuning of the data in order to impove the results.

## KNN Model

In [30]:
knn = KNeighborsClassifier(n_neighbors = 100)
knn.fit(X_train, y_train)

knn_preds = knn.predict_proba(X_valid)[:,1]

lr_valid_auc, lr_valid_accuracy, lr_valid_recall, \
    lr_valid_precision, lr_valid_specificity = print_report(y_valid,knn_preds, 0.5)

AUC:0.608
accuracy:0.592
recall:0.541
precision:0.148
specificity:0.557
prevalence:0.114
 


## Stochastic Gradient Descent Model

In [31]:
sgdc = SGDClassifier(loss='log_loss', alpha=0.1, random_state=42)
sgdc.fit(X_train, y_train)

sgd_preds = sgdc.predict_proba(X_valid)[:,1]

print('Stochastic Gradient Descent')
print('Validation:')
sgdc_valid_auc, sgdc_valid_accuracy, sgdc_valid_recall, \
                sgdc_valid_precision, sgdc_valid_specificity = print_report(y_valid,sgd_preds, 0.5)

Stochastic Gradient Descent
Validation:
AUC:0.629
accuracy:0.698
recall:0.435
precision:0.173
specificity:0.732
prevalence:0.114
 


## Decision Tree

In [32]:
dc_clf = DecisionTreeClassifier(random_state=42, max_depth = 10)
dc_clf.fit(X_train, y_train)

dc_preds_proba = dc_clf.predict_proba(X_valid)[:,1]
dc_preds = dc_clf.predict(X_valid)

lr_valid_auc, lr_valid_accuracy, lr_valid_recall, \
    lr_valid_precision, lr_valid_specificity = print_report(y_valid,dc_preds_proba, 0.5)

AUC:0.598
accuracy:0.584
recall:0.562
precision:0.149
specificity:0.583
prevalence:0.114
 


## Random Forest

In [33]:
rf_clf = RandomForestClassifier(random_state=111, max_depth = 6)

rf_clf.fit(X_train, y_train)

rf_preds = rf_clf.predict(X_valid)
rf_preds_proba = rf_clf.predict_proba(X_valid)[:, 1]

lr_valid_auc, lr_valid_accuracy, lr_valid_recall, \
    lr_valid_precision, lr_valid_specificity = print_report(y_valid,rf_preds_proba, 0.5)

AUC:0.628
accuracy:0.610
recall:0.562
precision:0.159
specificity:0.617
prevalence:0.114
 


## Linear SVC

In [34]:
lsvc_clf = LinearSVC(random_state=111)
lsvc_clf.fit(X_train, y_train)

lsvc_preds = lsvc_clf.decision_function(X_valid)

lr_valid_auc, lr_valid_accuracy, lr_valid_recall, \
    lr_valid_precision, lr_valid_specificity = print_report(y_valid,lsvc_preds, 0.5)

AUC:0.630
accuracy:0.879
recall:0.074
precision:0.359
specificity:0.983
prevalence:0.114
 


## Gradient Boosting Model

In [35]:
gb_clf = GradientBoostingClassifier(n_estimators = 100, criterion='friedman_mse', learning_rate = 1.0, max_depth = 3,\
                                    random_state = 111)

gb_clf.fit(X_train, y_train)
gb_preds = gb_clf.predict(X_valid)
gb_preds_proba = gb_clf.predict_proba(X_valid)[:, 1]

lr_valid_auc, lr_valid_accuracy, lr_valid_recall, \
    lr_valid_precision, lr_valid_specificity = print_report(y_valid,gb_preds_proba, 0.5)

AUC:0.602
accuracy:0.577
recall:0.578
precision:0.150
specificity:0.576
prevalence:0.114
 


## XGB Model

In [36]:
X_train_try1 = X_train.to_numpy()
X_valid_try1 = X_valid.to_numpy()

In [37]:
xgb_clf = xgb.XGBClassifier(max_depth=3, learning_rate = 1.0, use_label_encoder = False,\
                            eval_metric = 'logloss')
xgb_clf.fit(X_train_try1, y_train)

xgb_preds = xgb_clf.predict(X_valid_try1)
xgb_preds_proba = xgb_clf.predict_proba(X_valid_try1)[:, 1]

lr_valid_auc, lr_valid_accuracy, lr_valid_recall, \
    lr_valid_precision, lr_valid_specificity = print_report(y_valid,xgb_preds_proba, 0.5)

AUC:0.606
accuracy:0.581
recall:0.567
precision:0.149
specificity:0.582
prevalence:0.114
 


## Catboost Model

In [38]:
catb=CatBoostClassifier(iterations=200, depth=3, learning_rate=1.0, random_state = 111)
catb.fit(X_train, y_train)
catb_preds = catb.predict_proba(X_valid)[:, 1]

lr_valid_auc, lr_valid_accuracy, lr_valid_recall, \
    lr_valid_precision, lr_valid_specificity = print_report(y_valid,catb_preds, 0.5)

0:	learn: 0.6712826	total: 184ms	remaining: 36.7s
1:	learn: 0.6673275	total: 206ms	remaining: 20.4s
2:	learn: 0.6654723	total: 220ms	remaining: 14.4s
3:	learn: 0.6643218	total: 236ms	remaining: 11.6s
4:	learn: 0.6634768	total: 251ms	remaining: 9.78s
5:	learn: 0.6621311	total: 262ms	remaining: 8.47s
6:	learn: 0.6607321	total: 275ms	remaining: 7.58s
7:	learn: 0.6601515	total: 287ms	remaining: 6.89s
8:	learn: 0.6596016	total: 301ms	remaining: 6.39s
9:	learn: 0.6591848	total: 313ms	remaining: 5.94s
10:	learn: 0.6583812	total: 322ms	remaining: 5.54s
11:	learn: 0.6577374	total: 337ms	remaining: 5.28s
12:	learn: 0.6566526	total: 352ms	remaining: 5.06s
13:	learn: 0.6555355	total: 364ms	remaining: 4.84s
14:	learn: 0.6546320	total: 376ms	remaining: 4.64s
15:	learn: 0.6539351	total: 390ms	remaining: 4.49s
16:	learn: 0.6530831	total: 400ms	remaining: 4.31s
17:	learn: 0.6524436	total: 411ms	remaining: 4.16s
18:	learn: 0.6521580	total: 423ms	remaining: 4.03s
19:	learn: 0.6514755	total: 435ms	remaini

# Hyper Parameter Tuning

In [39]:
recall_scoring = make_scorer(recall_score)

## Decision Tree - Hyper Parameter Tuning

In [40]:
dc_grid = {'max_features':['auto','sqrt'], # maximum number of features to use at each split
           'max_depth':range(1,11,1), # maximum depth of the tree
           'min_samples_split':range(2,10,2), # minimum number of samples to split a node
           'criterion':['gini','entropy']} # criterion for evaluating a split

dc_random = RandomizedSearchCV(estimator = dc_clf, param_distributions = dc_grid, 
                               n_iter = 20, cv = 2, scoring=recall_scoring,
                               verbose = 1, random_state = 111)

dc_random.fit(X_train, y_train)

dc_random.best_params_

dc_hp_preds = dc_random.best_estimator_.predict(X_valid)
dc_hp_preds_proba = dc_random.best_estimator_.predict_proba(X_valid)[:,1]
roc_auc_score(y_valid, dc_hp_preds_proba)

Fitting 2 folds for each of 20 candidates, totalling 40 fits


0.5229444651667485

In [41]:
recall_score(y_valid, dc_hp_preds)

0.6217765042979942

## Random Forest - Hyper Parameter Tuning

In [43]:
rf_grid = {'n_estimators':range(200,1000,200), # number of trees
           'max_features':['auto','sqrt'], # maximum number of features to use at each split
           'max_depth':range(1,11,1), # maximum depth of the tree
           'min_samples_split':range(2,10,2), # minimum number of samples to split a node
           'criterion':['gini','entropy']} # criterion for evaluating a split

rf_random = RandomizedSearchCV(estimator = rf_clf, param_distributions = rf_grid, 
                               n_iter = 20, cv = 2, scoring=recall_scoring,
                               verbose = 1, random_state = 111)

rf_random.fit(X_train, y_train)

rf_random.best_params_

rf_hp_preds = rf_random.best_estimator_.predict(X_valid)
rf_hp_preds_proba = rf_random.best_estimator_.predict_proba(X_valid)[:,1]
roc_auc_score(y_valid, rf_hp_preds_proba)

Fitting 2 folds for each of 20 candidates, totalling 40 fits


0.6294584696766755

In [44]:
recall_score(y_valid, rf_hp_preds)

0.563323782234957

## XGBoost - Hyper Parameter Tuning

In [45]:
X_train_try2 = X_train.to_numpy()
X_valid_try2 = X_valid.to_numpy()

In [46]:
xgb_grid = params = {
        'min_child_weight': [1, 5, 8, 10],
        'gamma': [0.5, 1, 1.5, 2, 5],
        'subsample': [0.6, 0.8, 1.0],
        'colsample_bytree': [0.6, 0.8, 0.9, 1.0],
        'max_depth': [3, 4, 5]
        } # criterion for evaluating a split

xgb_random = GridSearchCV(estimator = xgb_clf, param_grid = xgb_grid,
                               cv = 2, scoring = recall_scoring,
                               verbose = 1)

xgb_random.fit(X_train_try2, y_train)

xgb_random.best_params_

xgb_hp_preds = xgb_random.best_estimator_.predict(X_valid_try2)
xgb_hp_preds_proba = xgb_random.best_estimator_.predict_proba(X_valid_try2)[:,1]
roc_auc_score(y_valid, xgb_hp_preds_proba)

Fitting 2 folds for each of 720 candidates, totalling 1440 fits


0.6199782769027314

In [102]:
recall_score(y_valid, xgb_hp_preds)

0.775

###  Hyper Parameter Tuning Results
* We can see that even after the hyper-parameter tuning XGB came out with the highest score of 0.775 beating out all the other models.